packages/flutter/lib/src/widgets/platform_menu_bar.dart

```diff
@@ -0,0 +1,886 @@
+    if (topLevelMenus.isNotEmpty) {
+      int index = 0;
+      for (final MenuItem childItem in topLevelMenus) {
+        if (childItem is! DefaultPlatformMenuDelegateSerializable) {
+          throw UnimplementedError(
+              'Tried to serialize a menu item that was not a DefaultPlatformMenuDelegateSerializer');
+        }
+        representation.addAll((childItem as DefaultPlatformMenuDelegateSerializable)
+            .toChannelRepresentation(this, index: index, count: topLevelMenus.length, getId: _getId));
+        index += 1;
+      }
+    }
+    channel.invokeMethod<void>(_kMenuSetMethod, representation);
+  }
+
+  /// Defines the channel that the [DefaultPlatformMenuDelegate] uses to
+  /// communicate with the platform.
+  ///
+  /// Defaults to [SystemChannels.menu].
+  final MethodChannel channel;
+
+  /// Get the next serialization ID.
+  ///
+  /// This is called by each DefaultPlatformMenuDelegateSerializer when
+  /// serializing a new object so that it has a unique ID.
+  int _getId(MenuItem item) {
+    _serial += 1;
+    _idMap[_serial] = item;
+    return _serial;
+  }
+
+  @override
+  bool debugLockDelegate(BuildContext context) {
+    assert(() {
+      // It's OK to lock if the lock isn't set, but not OK if a different
+      // context is locking it.
+      if (_lockedContext != null && _lockedContext != context) {
+        return false;
+      }
+      _lockedContext = context;
+      return true;
+    }());
+    return true;
+  }
+
+  @override
+  bool debugUnlockDelegate(BuildContext context) {
+    assert(() {
+      // It's OK to unlock if the lock isn't set, but not OK if a different
+      // context is unlocking it.
+      if (_lockedContext != null && _lockedContext != context) {
+        return false;
+      }
+      _lockedContext = null;
+      return true;
+    }());
+    return true;
+  }
+
+  // Handles the method calls from the plugin to forward to selection and
+  // open/close callbacks.
+  Future<void> _methodCallHandler(MethodCall call) async {
+    final int id = call.arguments as int;
+    assert(_idMap.containsKey(id),
+        'Received a menu ${call.method} for a menu item with an ID that was not recognized: $id');
+    if (!_idMap.containsKey(id)) {
+      return;
+    }
+    final MenuItem item = _idMap[id]!;
+    if (item is PlatformMenuItem && call.method == _kMenuSelectedCallbackMethod) {
+      item.onSelected?.call();
+    } else if (item is PlatformMenu && call.method == _kMenuItemOpenedMethod) {
+      item.onOpen?.call();
+    } else if (item is PlatformMenu && call.method == _kMenuItemClosedMethod) {
+      item.onClose?.call();
+    }
+    return;
+  }
+}
+
+/// A menu bar that uses the platform's native APIs to construct and render a
+/// menu described by a [PlatformMenu]/[PlatformMenuItem] hierarchy.
+///
+/// This widget is especially useful on macOS, where a system menu is a required
+/// part of every application. Flutter only includes support for macOS out of
```

Comment:
Somewhere here maybe include a note that other platforms usually render their menu as part of the app window, which flutter also supports using x.

(Right now it sounds a little bit like the missing support on other platforms is a big deal, when it really isn't)